### PR TITLE
Sai999 python kinesis

### PR DESCRIPTION
**Pull request by meetreks**
*Issue #, if available:*

*Description of changes:*

This PR is to cater for unit & integration mock tests involving kinesis-lambda-dynamodb.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by wtromano-aws** [comment]
wtromano-aws review in progress.

File: python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        """
+        
+        test_event = self.load_test_event("sample_test_event")
+        from src.app import lambda_handler
+        test_return = lambda_handler(event=test_event,context=None)
+        
+        self.assertEqual( test_return["statusCode"] , 200)
+        self.assertEqual( test_return["body"] , "Kinesis events processed and persisted to DynamoDB table")
+        
+        #check the mocked DynamoDB to see if PK matches to what we passed to the Kinesss Event
```
Comment (wtromano-aws) [inline python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py:95]: Spelling - Kinesis

File: python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py
Comment (wtromano-aws) [inline python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py:1]: pylint:

mock_test.py:27:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:28:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:41:0: C0301: Line too long (108/100) (line-too-long)
mock_test.py:49:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:50:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:55:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:72:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:78:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:87:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:91:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:93:0: C0301: Line too long (107/100) (line-too-long)
mock_test.py:94:0: C0303: Trailing whitespace (trailing-whitespace)
mock_test.py:96:0: C0301: Line too long (139/100) (line-too-long)
mock_test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
mock_test.py:14:0: E0401: Unable to import 'moto' (import-error)
mock_test.py:62:13: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
mock_test.py:62:43: C0103: Variable name "fp" doesn't conform to snake_case naming style (invalid-name)
mock_test.py:73:17: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
mock_test.py:73:61: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
mock_test.py:66:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
mock_test.py:89:8: C0415: Import outside toplevel (src.app.lambda_handler) (import-outside-toplevel)
mock_test.py:6:0: W0611: Unused datetime imported from datetime (unused-import)
mock_test.py:9:0: W0611: Unused uuid4 imported from uuid (unused-import)
mock_test.py:12:0: W0611: Unused Session imported from boto3 (unused-import)

**Review by wtromano-aws — CHANGES_REQUESTED** [review]
Overall, I like the approach and the way it is coded, this will be a great example - thank you!
But some issues prevented it from working for me, so I am not able to fully test it either with unit or integration.  Let me know when the issues are addressed as commented, and I will try again.

File: python-test-samples/README.md
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/README.md:1]: Yes, you are correct. Restored the latest one from main and added my new project details.

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  - [Sample project description](#sample-project-description)
+  - [Run the Unit Test](#run-the-unit-test)
+  - [Run the Integration Test](#run-the-integration-test)
+---
+
+## Key Files in the Project
+  - [app.py](src/app.py) - Lambda handler code to test
+  - [template.yaml](template.yaml) - SAM script for deployment
+  - [mock_test.py](tests/unit/mock_test.py) - Unit test using mocks
+  - [test_kinesis.py](tests/integration/kinesis.py) - Integration tests on a live stack
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:32]: Corrected the broken link.

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  
+[Top](#contents)
+
+---
+
+## About this Pattern
+
+### System Under Test (SUT)
+
+The SUT is a streaming data processing system. A Lambda function has an Event Source Mapping to a Kinesis Data Stream. The Lambda Event Source Mapping(ESM) polls the Kinesis Data Stream and then synchronously invokes the Lambda function with a batch of messages. The Lambda function processes batches of messages and writes results to a DynamoDB Table.
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:42]: Done as requested.

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+![System Under Test (SUT)](img/system-under-test.png)
+
+### Goal
+
+The goal of this example is to show how to test Lambda functions that are part of a streaming data processing application. In streaming workloads the number of messages that are sent to Lambda in a batch can change with the rate of messages being published to the stream, so we show testing with different sized batches.
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:48]: Done

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+This example contains a sample event with pre-generated records but we can also use the SAM CLI to generate a test event that matches the shape of the event that will be received by a Lambda functions subscribed to a Kinesis Data Stream. Another option to get test events is to log them, when logging is set to debug, in the Lambda function. This allows you to capture real world messages to use for testing, just ensure that they don’t contain any sensitive data.
+
+When you configure the Lambda ESM for Kinesis you specify a batch size. This batch size functions as a maximum batch size. Your Lambda function may still be invoked with batches smaller than this, but not larger. Create a test events that contain a single message, your expected average message batch, and your max batch size. If the size of the individual records varies by a large amount you should also create some test payloads that have batches of larger records.
+
+![Unit Test Description](img/unit-test-description.png)
+
+
+### Run the Unit Test
+
+[mock_test.py](src/tests/unit/mock_test.py)
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:109]: corrected

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+In the [unit test](src/tests/unit/test-handler.test.ts#L44), all references and calls to the DynamoDB service are mocked using aws-sdk-client-mock client.
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:111]: Removed the line reference and made it general.

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+To run the unit tests
+``` shell
+# Create and Activate a Python Virtual Environment
+# One-time setup
+kinesis-lambda-dynamodb$ pip3 install virtualenv
+kinesis-lambda-dynamodb$ python3 -m venv venv
+kinesis-lambda-dynamodb$ source ./venv/bin/activate
+
+# install dependencies
+kinesis-lambda-dynamodb$ pip3 install -r tests/requirements.txt
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:121]: Added to use a pip parameter --use-pep517 when installing the package.

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+# run unit tests with mocks
+kinesis-lambda-dynamodb$ python3 -m pytest -s tests/unit  -v
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:124]: Hi I am running Pytest and it passes the tests. I am not sure why the tests are failing in your environment?

File: python-test-samples/kinesis-lambda-dynamodb/img/diagram.png
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/img/diagram.png:1]: revamped it to match the TypeScript example

File: python-test-samples/kinesis-lambda-dynamodb/img/system-under-test-description.png
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/img/system-under-test-description.png:1]: Kept it in sync with the typescript examples.

File: python-test-samples/kinesis-lambda-dynamodb/img/system-under-test.png
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/img/system-under-test.png:1]: Kept it in sync with Type script example.

File: python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py (hunk 1 above)
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py:95]: Done

File: python-test-samples/kinesis-lambda-dynamodb/README.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+## Integration Test
+
+### Integration Test description
+
+In order to run integration tests in the cloud we will use an event listener pattern to capture the items processed in the SUT. While we could just look directly at the DynamoDB table in the SUT, using a second table allows us to capture timings and other metadata, it also makes this pattern more re-usable across different types of streaming applications. This event listener should only be deployed in non-production environments for the purposes of testing. In Step 1 in the diagram below the test establishes a long polling pattern with DynamoDB. In Step 2, the test uses the Kinesis Data Streams APIs to send messages into the SUT. The streaming application processes those messages and in Step 3 the Event Listener Lambda function receives the messages and writes them to the Event Listener DynamoDB table. The test receives the results via the long polling mechanism and examines the results.
+
+![Integration Test Description](img/integration-test-description.png)
+
+### Run the Integration Tests
+[test_kinesis.py](tests/integration/test_api_gateway.py)
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:141]: Fixed

File: python-test-samples/kinesis-lambda-dynamodb/img/integration-test-description.png
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/img/integration-test-description.png:1]: Used the same images in the Typescript example to keep it in sync.

File: python-test-samples/kinesis-lambda-dynamodb/src/app.py
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/src/app.py:1]: Fixed linting issues

File: python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/tests/unit/mock_test.py:1]: Fixed

File: python-test-samples/kinesis-lambda-dynamodb/tests/integration/test_kinesis.py
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/tests/integration/test_kinesis.py:1]: Fixed

File: python-test-samples/kinesis-lambda-dynamodb/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/README.md:124]: image

File: python-test-samples/kinesis-lambda-dynamodb/src/app.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        
+    return {
+        "statusCode": 200,
+        "body": "Kinesis events processed and persisted to DynamoDB table"
+    }
+    
+def write_to_dynamodb(items):
+    with dynamodb_table.batch_writer() as batch:
+        for item in items:
+            batch.put_item(Item=item)
```
[the remark this answers is not in the document]
Comment (meetreks) [reply python-test-samples/kinesis-lambda-dynamodb/src/app.py]: Kept it in sync with the Typescript Example

**Review by wtromano-aws — COMMENTED** [review]
Integration tests are failing, see comment on template.yaml for issue.

**Comment by wtromano-aws** [comment]
Code verified as functional and comments resolved.  Thank you so much for all the efforts, this will be a very helpful pattern!

**Review by wtromano-aws — APPROVED** [review]

**Pull request merged**